File: scripts/generate-badges.mjs
#!/usr/bin/env node
/**
 * Generates SEA-V hex badges — Strava-style typography inside, page-colored
 * rings. Two text tiers only (hero + sub) — the category "tag" chip was
 * dropped because it's already shown as normal HTML text next to every badge
 * image (achievements.js's trophy tiles, the unlock modal, the recent feed),
 * so it was pure duplication and it was eating room that made the hero/sub
 * text shrink more than it needed to. Full words only now — no more
 * mid-word truncations like "COMMER" or "HELO" forced by cramped space.
 * Run: node scripts/generate-badges.mjs
 */
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const OUT = path.join(__dirname, "../img/badges");
const PAGE_COLORS_PATH = path.join(__dirname, "page-colors.json");
const PAGE_COLORS_JS = path.join(__dirname, "../js/seav-page-colors.js");
const BADGE_COPY_PATH = path.join(__dirname, "badge-copy.json");

const PAGES = JSON.parse(fs.readFileSync(PAGE_COLORS_PATH, "utf8"));
const BADGE_DEFS = JSON.parse(fs.readFileSync(BADGE_COPY_PATH, "utf8"));

const CX = 60;
const CY = 56;
const TEXT_CY = 56;
const MAX_TEXT_WIDTH = 54;
const TEXT = "#0F172A";
const FONT =
  "system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,sans-serif";

function pageRing(page) {
  const [a, b, c] = page.ring;
  return { a, b, c };
}

function pageInner(page) {
  const [a, b, c] = page.ring;
  return { fill: [b, c], pill: [a, c] };
}

function esc(text) {
  return String(text)
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;");
}

function estimateWidth(text, fontSize, tracking = 0.4) {
  return text.length * (fontSize * 0.55 + tracking);
}

function fitFontSize(text, maxWidth, startSize, minSize = 8) {
  let size = startSize;
  while (size > minSize) {
    if (estimateWidth(text, size) <= maxWidth) return Math.round(size * 10) / 10;
    size -= 0.5;
  }
  return minSize;
}

// Only wraps if a line genuinely can't fit even at the minimum size — full
// words are kept together whenever possible instead of being abbreviated.
function wrapLines(text, maxCharsPerLine = 11) {
  const value = String(text).trim();
  if (!value) return [""];
  const words = value.split(" ");
  if (words.length === 1 || value.length <= maxCharsPerLine) return [value];

  const lines = [];
  let current = words[0];
  for (const word of words.slice(1)) {
    const candidate = `${current} ${word}`;
    if (candidate.length <= maxCharsPerLine) current = candidate;
    else {
      lines.push(current);
      current = word;
    }
  }
  lines.push(current);
  return lines.slice(0, 2);
}

function heroBaseSize(hero) {
  const h = String(hero);
  if (/^\d/.test(h)) {
    if (/m/i.test(h)) return 22;
    if (h.length <= 2) return 28;
    if (h.length <= 4) return 22;
    return 18;
  }
  if (h.length <= 5) return 17;
  if (h.length <= 8) return 14;
  return 12;
}

function heroTracking(hero) {
  const h = String(hero);
  if (h.length > 8) return 0.15;
  if (h.length > 5) return 0.3;
  if (/^\d/.test(h)) return 0.8;
  return 0.5;
}

function renderTypography({ hero, sub, theme }) {
  const heroLines = wrapLines(hero, 9);
  const subLines = wrapLines(sub, 12);

  let heroSize = heroBaseSize(heroLines[0]);
  for (const line of heroLines) {
    heroSize = Math.min(heroSize, fitFontSize(line, MAX_TEXT_WIDTH, heroSize, 9));
  }

  let subSize = 7.4;
  for (const line of subLines) {
    subSize = Math.min(subSize, fitFontSize(line, MAX_TEXT_WIDTH, 7.4, 6.4));
  }

  let heroLineH = Math.max(heroSize * 0.85, 9);
  let subLineH = 8.6;
  const gap = 4;

  let blockH = heroLines.length * heroLineH + gap + subLines.length * subLineH;
  const maxBlock = 40;
  if (blockH > maxBlock) {
    const scale = maxBlock / blockH;
    heroSize = Math.max(9, Math.round(heroSize * scale * 10) / 10);
    subSize = Math.max(6.4, Math.round(subSize * scale * 10) / 10);
    heroLineH = Math.max(heroSize * 0.85, 8);
    subLineH = 7.8;
    blockH = heroLines.length * heroLineH + gap + subLines.length * subLineH;
  }

  const topY = TEXT_CY - blockH / 2;
  const heroStartY = topY + heroSize * 0.78;
  const subStartY = heroStartY + (heroLines.length - 1) * heroLineH + gap + subSize * 0.72;
  const accentY = Math.min(subStartY + (subLines.length - 1) * subLineH + 4, 71.5);

  const heroTspans = heroLines
    .map(
      (line, i) =>
        `<tspan x="60" ${i === 0 ? `y="${heroStartY.toFixed(1)}"` : ""} dy="${i === 0 ? 0 : heroLineH.toFixed(1)}">${esc(line)}</tspan>`
    )
    .join("");

  const subTspans = subLines
    .map(
      (line, i) =>
        `<tspan x="60" ${i === 0 ? `y="${subStartY.toFixed(1)}"` : ""} dy="${i === 0 ? 0 : subLineH.toFixed(1)}">${esc(line)}</tspan>`
    )
    .join("");

  return `
  <g font-family="${FONT}" text-anchor="middle">
    <text font-size="${heroSize}" font-weight="900" letter-spacing="${heroTracking(hero)}" fill="${TEXT}">${heroTspans}</text>
    <text font-size="${subSize}" font-weight="700" letter-spacing="0.6" fill="${TEXT}">${subTspans}</text>
    <line x1="40" y1="${accentY.toFixed(1)}" x2="80" y2="${accentY.toFixed(1)}" stroke="${theme.sidebar}" stroke-width="1.6" stroke-linecap="round" opacity="0.4"/>
  </g>`;
}

function hexPoints(cx, cy, r, startDeg = -90) {
  return Array.from({ length: 6 }, (_, i) => {
    const a = ((startDeg + i * 60) * Math.PI) / 180;
    return [cx + r * Math.cos(a), cy + r * Math.sin(a)];
  });
}

function hexPath(cx, cy, r, startDeg = -90) {
  const pts = hexPoints(cx, cy, r, startDeg);
  return `M ${pts.map((p) => p.map((n) => n.toFixed(2)).join(" ")).join(" L ")} Z`;
}

function vertexTicks(uid) {
  const outer = hexPoints(CX, CY, 48);
  const inner = hexPoints(CX, CY, 44);
  return outer
    .map(
      (p, i) =>
        `<line x1="${inner[i][0].toFixed(2)}" y1="${inner[i][1].toFixed(2)}" x2="${p[0].toFixed(2)}" y2="${p[1].toFixed(2)}" stroke="url(#ring-${uid})" stroke-width="2.4" stroke-linecap="round"/>`
    )
    .join("\n    ");
}

function buildSvg({ file, tier, hero, sub, page, locked = false }) {
  const theme = PAGES[page] || PAGES.seatime;
  const ring = pageRing(theme);
  const inner = pageInner(theme);
  const uid = Math.random().toString(36).slice(2, 8);
  const outerHex = hexPath(CX, CY, 50);
  const innerHex = hexPath(CX, CY, 40);
  const clip = hexPath(CX, CY, 40);
  const textClip = hexPath(CX, CY, 38);

  const lockOverlay = locked
    ? `
  <path d="${innerHex}" fill="#F1F5F9" opacity="0.82"/>
  <text x="60" y="56" text-anchor="middle" font-family="${FONT}" font-size="10" font-weight="800" letter-spacing="1.2" fill="#64748B">LOCKED</text>`
    : "";

  return `<?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 120 120" role="img" aria-label="${esc(`${hero} ${sub}`)}" data-source-page="${page}" data-tier="${tier}">
  <defs>
    <linearGradient id="ring-${uid}" x1="0%" y1="0%" x2="100%" y2="100%">
      <stop offset="0%" stop-color="${ring.a}"/>
      <stop offset="50%" stop-color="${ring.b}"/>
      <stop offset="100%" stop-color="${ring.c}"/>
    </linearGradient>
    <radialGradient id="disc-${uid}" cx="38%" cy="32%" r="68%">
      <stop offset="0%" stop-color="${inner.fill[1]}"/>
      <stop offset="100%" stop-color="${inner.fill[0]}"/>
    </radialGradient>
    <linearGradient id="shine-${uid}" x1="0%" y1="0%" x2="0%" y2="100%">
      <stop offset="0%" stop-color="#FFFFFF" stop-opacity="0.24"/>
      <stop offset="50%" stop-color="#FFFFFF" stop-opacity="0"/>
    </linearGradient>
    <clipPath id="txt-${uid}">
      <path d="${textClip}"/>
    </clipPath>
    <filter id="sh-${uid}" x="-20%" y="-15%" width="140%" height="150%">
      <feDropShadow dx="0" dy="5" stdDeviation="4" flood-color="#0F172A" flood-opacity="0.28"/>
    </filter>
  </defs>

  <g filter="url(#sh-${uid})">
    <path d="${outerHex}" fill="none" stroke="url(#ring-${uid})" stroke-width="5" stroke-linejoin="round"/>
    ${vertexTicks(uid)}
    <path d="${clip}" fill="url(#disc-${uid})"/>
    <path d="${clip}" fill="url(#shine-${uid})"/>
    <path d="${innerHex}" fill="none" stroke="#FFFFFF" stroke-opacity="0.16" stroke-width="1"/>
  </g>

  <g clip-path="url(#txt-${uid})">
  ${renderTypography({ hero, sub, theme })}
  </g>

  ${lockOverlay}
</svg>`;
}

function syncPageColorsJs() {
  const payload = { pages: PAGES };
  const body = `/* Auto-generated from scripts/page-colors.json — run scripts/generate-badges.mjs */\nwindow.SeavPageColors = ${JSON.stringify(payload, null, 2)};\n`;
  fs.writeFileSync(PAGE_COLORS_JS, body);
}

fs.mkdirSync(OUT, { recursive: true });
syncPageColorsJs();

for (const def of BADGE_DEFS) {
  fs.writeFileSync(path.join(OUT, `${def.file}.svg`), buildSvg(def));
}

fs.writeFileSync(
  path.join(OUT, "locked.svg"),
  buildSvg({
    file: "locked",
    tier: "silver",
    hero: "LOCKED",
    sub: "MILESTONE",
    page: "achievements",
    locked: true
  })
);

fs.writeFileSync(
  path.join(OUT, "default.svg"),
  buildSvg({
    file: "default",
    tier: "default",
    hero: "SEA-V",
    sub: "CREW BADGE",
    page: "achievements"
  })
);

console.log(`Generated ${BADGE_DEFS.length + 2} typographic hex badges in ${OUT}`);

File: scripts/generate-badges.py
#!/usr/bin/env python3
"""Generate flat-top hex badges. Mirrors scripts/generate-badges.mjs."""
import math
import os
import random
import string

OUT = os.path.join(os.path.dirname(__file__), "..", "img", "badges")
CX, CY = 60, 56

TIERS = {
    "bronze": {"a": "#FDBA74", "b": "#C2410C", "c": "#FED7AA"},
    "silver": {"a": "#E2E8F0", "b": "#64748B", "c": "#F8FAFC"},
    "gold": {"a": "#FDE68A", "b": "#D97706", "c": "#FFFBEB"},
    "platinum": {"a": "#67E8F9", "b": "#0891B2", "c": "#ECFEFF"},
    "default": {"a": "#93C5FD", "b": "#2563EB", "c": "#EFF6FF"},
}

ICON_OUTLINE = "#0F172A"

CATEGORIES = {
    "sea": {"inner": ["#0C4A6E", "#0369A1"], "accent": "#FFFFFF", "glow": "#38BDF8"},
    "vessel": {"inner": ["#7C2D12", "#C2410C"], "accent": "#FFFFFF", "glow": "#FB923C"},
    "passage": {"inner": ["#581C87", "#7E22CE"], "accent": "#FFFFFF", "glow": "#C084FC"},
    "ocean": {"inner": ["#1E3A8A", "#1D4ED8"], "accent": "#FFFFFF", "glow": "#60A5FA"},
    "polar": {"inner": ["#164E63", "#0891B2"], "accent": "#FFFFFF", "glow": "#67E8F9"},
    "watch": {"inner": ["#312E81", "#4338CA"], "accent": "#FFFFFF", "glow": "#818CF8"},
    "career": {"inner": ["#14532D", "#15803D"], "accent": "#FFFFFF", "glow": "#4ADE80"},
    "ops": {"inner": ["#713F12", "#B45309"], "accent": "#FFFFFF", "glow": "#FBBF24"},
    "helm": {"inner": ["#0F2744", "#D97706"], "accent": "#FFFFFF", "glow": "#FBBF24"},
}

ICONS = {
    "sea": """
    <path d="M34 52c6-4 12-4 18 0s12 4 18 0 12-4 18 0" stroke="currentColor" stroke-width="2.8" fill="none" stroke-linecap="round"/>
    <path d="M30 58c6-3 12-3 18 0s12 3 18 0 12-3 18 0" stroke="currentColor" stroke-width="2.2" fill="none" opacity="0.7" stroke-linecap="round"/>
    <circle cx="60" cy="42" r="11" stroke="currentColor" stroke-width="2.6" fill="none"/>
    <path d="M60 35v7l5 3" stroke="currentColor" stroke-width="2.4" stroke-linecap="round" fill="none"/>""",
    "vessel": """
    <path d="M36 56h48l-7-16H43l-7 16z" fill="currentColor"/>
    <path d="M30 59c9-3 18-3 27 0s18 3 27 0" stroke="currentColor" stroke-width="2.4" fill="none" stroke-linecap="round"/>
    <path d="M50 40h20v10H50z" fill="currentColor" opacity="0.55"/>
    <path d="M54 35h12v5H54z" fill="currentColor"/>""",
    "passage": """
    <circle cx="40" cy="46" r="5" fill="currentColor"/>
    <circle cx="80" cy="40" r="5" fill="currentColor"/>
    <path d="M45 45c14-9 28-9 34-6" stroke="currentColor" stroke-width="2.6" stroke-dasharray="5 3" fill="none" stroke-linecap="round"/>
    <path d="M52 54l7-9 7 6 9-11" stroke="currentColor" stroke-width="2.2" fill="none" stroke-linecap="round" stroke-linejoin="round"/>""",
    "ocean": """
    <circle cx="60" cy="48" r="20" stroke="currentColor" stroke-width="2.4" fill="none" opacity="0.45"/>
    <path d="M40 48c7-11 16-11 20 0s16 11 20 0" stroke="currentColor" stroke-width="2.6" fill="none" stroke-linecap="round"/>
    <path d="M46 58h28l-5-10H51l-5 10z" fill="currentColor"/>""",
    "polar": """
    <path d="M60 32l5 14h14l-11 8 4 14-12-8-12 8 4-14-11-8h14z" fill="currentColor"/>
    <path d="M42 60h36" stroke="currentColor" stroke-width="2" opacity="0.5"/>""",
    "watch": """
    <rect x="40" y="36" width="40" height="24" rx="3" stroke="currentColor" stroke-width="2.6" fill="none"/>
    <path d="M44 52h32" stroke="currentColor" stroke-width="2" opacity="0.55"/>
    <circle cx="60" cy="48" r="7" stroke="currentColor" stroke-width="2.2" fill="none"/>
    <path d="M60 44v5l3 2" stroke="currentColor" stroke-width="2" stroke-linecap="round" fill="none"/>""",
    "helm": f"""
    <g transform="translate(60,48)" shape-rendering="geometricPrecision">
      <circle r="28" fill="none" stroke="currentColor" stroke-width="3.5"/>
      <circle r="22" fill="none" stroke="currentColor" stroke-width="2.2" opacity="0.65"/>
      <g stroke="currentColor" stroke-width="3.2" fill="none">
        <line x1="0" y1="-24" x2="0" y2="24"/>
        <line x1="-24" y1="0" x2="24" y2="0"/>
        <line x1="-17" y1="-17" x2="17" y2="17"/>
        <line x1="17" y1="-17" x2="-17" y2="17"/>
      </g>
      <g fill="currentColor" stroke="currentColor" stroke-width="1.2">
        <circle cy="-28" r="3"/><circle cy="28" r="3"/>
        <circle cx="-28" r="3"/><circle cx="28" r="3"/>
        <circle cx="-20" cy="-20" r="3"/><circle cx="20" cy="-20" r="3"/>
        <circle cx="-20" cy="20" r="3"/><circle cx="20" cy="20" r="3"/>
      </g>
      <circle r="7" fill="{ICON_OUTLINE}" stroke="currentColor" stroke-width="2.4"/>
      <circle r="3" fill="currentColor"/>
    </g>""",
    "career": """
    <path d="M46 36h28v7H46z" fill="currentColor" opacity="0.65"/>
    <path d="M42 43h36v7H42z" fill="currentColor" opacity="0.85"/>
    <path d="M38 50h44v7H38z" fill="currentColor"/>
    <path d="M52 57h16v5H52z" fill="currentColor"/>""",
    "ops": """
    <circle cx="46" cy="48" r="9" stroke="currentColor" stroke-width="2.6" fill="none"/>
    <circle cx="74" cy="48" r="9" stroke="currentColor" stroke-width="2.6" fill="none"/>
    <path d="M53 48h14" stroke="currentColor" stroke-width="3" stroke-linecap="round"/>
    <path d="M60 58v7M50 63h20" stroke="currentColor" stroke-width="2.4" stroke-linecap="round"/>""",
}

BADGE_DEFS = [
    ("sea-30-days", "default", "30", "DAYS AT SEA", "sea", False),
    ("sea-100-days", "default", "100", "DAYS AT SEA", "sea", False),
    ("sea-250-days", "silver", "250", "DAYS AT SEA", "sea", False),
    ("sea-500-days", "gold", "500", "DAYS AT SEA", "sea", False),
    ("sea-1-year", "gold", "1 YR", "SEA TIME", "sea", False),
    ("sea-3-years", "platinum", "3 YR", "SEA TIME", "sea", False),
    ("first-vessel-logged", "bronze", "1ST", "VESSEL", "vessel", False),
    ("vessels-3-served", "silver", "3", "YACHTS", "vessel", False),
    ("vessel-types-5", "gold", "5", "HULL TYPES", "vessel", False),
    ("large-yacht-50m", "silver", "50m+", "SUPERYACHT", "vessel", False),
    ("explorer-vessel", "silver", "EXP", "EXPLORER", "vessel", False),
    ("commercial-vessel", "silver", "COM", "COMMERCIAL", "vessel", False),
    ("offshore-100nm", "bronze", "100", "OFFSHORE NM", "passage", False),
    ("passage-500nm", "silver", "500", "PASSAGE NM", "passage", False),
    ("passage-1000nm", "gold", "1K", "PASSAGE NM", "passage", False),
    ("atlantic-crossing", "gold", "ATL", "CROSSING", "ocean", False),
    ("pacific-crossing", "platinum", "PAC", "CROSSING", "ocean", False),
    ("polar-navigation", "platinum", "POLAR", "NAVIGATION", "polar", False),
    ("first-watchkeeping", "bronze", "1ST", "WATCH", "watch", False),
    ("watchkeeping-100-days", "gold", "100", "BRIDGE DAYS", "watch", False),
    ("oow-level", "gold", "OOW", "DECK OFFICER", "watch", False),
    ("bridge-leader", "platinum", "LEAD", "BRIDGE LEADER", "helm", True),
    ("first-promotion", "silver", "UP", "PROMOTION", "career", False),
    ("senior-crew", "gold", "SNR", "SENIOR CREW", "career", False),
    ("officer-rank", "gold", "OFC", "OFFICER", "career", False),
    ("command-experience", "platinum", "CMD", "COMMAND", "career", False),
    ("tender-operations", "silver", "TDR", "TENDER OPS", "ops", False),
    ("watersports-operations", "silver", "H2O", "WATERSPORTS", "ops", False),
    ("crane-operations", "gold", "LIFT", "CRANE OPS", "ops", False),
    ("helicopter-operations", "platinum", "HELO", "FLIGHT OPS", "ops", False),
]


def uid():
    return "".join(random.choices(string.ascii_lowercase + string.digits, k=6))


def hex_points(cx, cy, r, start_deg=-90):
    pts = []
    for i in range(6):
        a = math.radians(start_deg + i * 60)
        pts.append((cx + r * math.cos(a), cy + r * math.sin(a)))
    return pts


def hex_path(cx, cy, r, start_deg=-90):
    pts = hex_points(cx, cy, r, start_deg)
    return "M " + " L ".join(f"{x:.2f} {y:.2f}" for x, y in pts) + " Z"


def main_font_size(main):
    n = len(main)
    if n >= 5:
        return 11
    if n == 4:
        return 12
    if n == 3:
        return 13
    return 15


def label_font_size(label):
    return 5.2 if len(label) > 12 else 5.8


def vertex_ticks(u):
    outer = hex_points(CX, CY, 48)
    inner = hex_points(CX, CY, 44)
    lines = []
    for i in range(6):
        lines.append(
            f'    <line x1="{inner[i][0]:.2f}" y1="{inner[i][1]:.2f}" '
            f'x2="{outer[i][0]:.2f}" y2="{outer[i][1]:.2f}" '
            f'stroke="url(#ring-{u})" stroke-width="2.4" stroke-linecap="round"/>'
        )
    return "\n".join(lines)


def inner_fill(u, c, split):
    clip = hex_path(CX, CY, 40)
    if split:
        return f"""
    <clipPath id="hex-inner-{u}"><path d="{clip}"/></clipPath>
    <g clip-path="url(#hex-inner-{u})">
      <rect x="0" y="0" width="{CX}" height="120" fill="{c['inner'][0]}"/>
      <rect x="{CX}" y="0" width="{CX}" height="120" fill="{c['inner'][1]}"/>
    </g>"""
    return f"""
    <path d="{clip}" fill="url(#disc-{u})"/>
    <path d="{clip}" fill="url(#shine-{u})"/>"""


def render_icon(icon, accent="#FFFFFF"):
    return f"""
  <g shape-rendering="geometricPrecision">
    <g color="{ICON_OUTLINE}" opacity="0.38" transform="translate(0, 2)">{icon}</g>
    <g color="{accent}" stroke="{ICON_OUTLINE}" stroke-width="2" stroke-linejoin="round" stroke-linecap="round" paint-order="stroke fill">{icon}</g>
  </g>"""


def build_svg(tier, main, label, cat, split=False, locked=False):
    t = TIERS.get(tier, TIERS["default"])
    c = CATEGORIES.get(cat, CATEGORIES["sea"])
    u = uid()
    icon = ICONS.get(cat, ICONS["sea"])
    main_size = main_font_size(main)
    label_size = label_font_size(label)
    outer_hex = hex_path(CX, CY, 50)
    inner_hex = hex_path(CX, CY, 40)
    pill_fill = c["inner"][0] if split else c["glow"]
    pill_text = "#FDE68A" if split else "#0F172A"
    lock = ""
    if locked:
        lock = f"""
  <path d="{inner_hex}" fill="#F1F5F9" opacity="0.78"/>
  <g transform="translate(60,54)" stroke="#64748B" fill="none" stroke-width="2.2" stroke-linecap="round">
    <rect x="-11" y="-4" width="22" height="18" rx="4" fill="#E2E8F0" stroke="#64748B"/>
    <path d="M-6 6v6a6 6 0 0 0 12 0v-6"/>
    <path d="M-14 -2h28"/>
  </g>"""
    return f"""<?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 120 120" role="img">
  <defs>
    <linearGradient id="ring-{u}" x1="0%" y1="0%" x2="100%" y2="100%">
      <stop offset="0%" stop-color="{t['a']}"/>
      <stop offset="50%" stop-color="{t['b']}"/>
      <stop offset="100%" stop-color="{t['c']}"/>
    </linearGradient>
    <radialGradient id="disc-{u}" cx="38%" cy="32%" r="68%">
      <stop offset="0%" stop-color="{c['inner'][1]}"/>
      <stop offset="100%" stop-color="{c['inner'][0]}"/>
    </radialGradient>
    <linearGradient id="shine-{u}" x1="0%" y1="0%" x2="0%" y2="100%">
      <stop offset="0%" stop-color="#FFFFFF" stop-opacity="0.24"/>
      <stop offset="50%" stop-color="#FFFFFF" stop-opacity="0"/>
    </linearGradient>
    <filter id="sh-{u}" x="-20%" y="-15%" width="140%" height="150%">
      <feDropShadow dx="0" dy="5" stdDeviation="4" flood-color="#0F172A" flood-opacity="0.28"/>
    </filter>
  </defs>

  <g filter="url(#sh-{u})">
    <path d="{outer_hex}" fill="none" stroke="url(#ring-{u})" stroke-width="5" stroke-linejoin="round"/>
{vertex_ticks(u)}
{inner_fill(u, c, split)}
    <path d="{inner_hex}" fill="none" stroke="#FFFFFF" stroke-opacity="0.14" stroke-width="1"/>
  </g>

{render_icon(icon, c['accent'])}

  <rect x="26" y="82" width="68" height="18" rx="4" fill="#0F172A" opacity="0.85"/>
  <rect x="27" y="83" width="66" height="16" rx="3" fill="{pill_fill}" opacity="0.96"/>
  <text x="60" y="94.5" text-anchor="middle"
    font-family="system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif"
    font-size="{main_size}" font-weight="800" letter-spacing="0.6"
    fill="{pill_text}">{main}</text>

  <text x="60" y="108" text-anchor="middle"
    font-family="system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif"
    font-size="{label_size}" font-weight="700" letter-spacing="0.9"
    fill="#64748B">{label}</text>

  {lock}
</svg>"""


def main():
    os.makedirs(OUT, exist_ok=True)
    for file, tier, main, label, cat, split in BADGE_DEFS:
        with open(os.path.join(OUT, f"{file}.svg"), "w", encoding="utf-8") as f:
            f.write(build_svg(tier, main, label, cat, split))
    with open(os.path.join(OUT, "locked.svg"), "w", encoding="utf-8") as f:
        f.write(build_svg("silver", "—", "LOCKED", "sea", locked=True))
    with open(os.path.join(OUT, "default.svg"), "w", encoding="utf-8") as f:
        f.write(build_svg("default", "SV", "SEA-V CREW", "sea"))
    print(f"Generated {len(BADGE_DEFS) + 2} hex badges in {OUT}")


if __name__ == "__main__":
    main()
